tla-types/src/main/scala/at/forsyte/apalache/tla/types/package.scala

```diff
@@ -1,22 +1,134 @@
+ * }}}
+ *
+ * =Exceptions=
+ *
+ * There are three categories of exceptions, which you may encounter when using [[tla]]:
+ *
+ * <ul>
+ *
+ * <li> [[typecomp.TBuilderTypeException TBuilderTypeException]]: This exception indicates that one or more arguments
+ * have incorrect types, w.r.t. the operator. For example, attempting to construct `1 + "a"` would yield this exception:
+ * {{{
+ *   val ex: TlaEx = tla.plus(tla.int(1), tla.str("a")) // throws TBuilderTypeException
+ * }}}
+ * as `+` requires two arguments of type `Int`. </li>
+ *
+ * <li> [[IllegalArgumentException]]: This exception indicates that one or more arguments has an illegal shape, w.r.t.
+ * the operator, or that the number of arguments is incorrect. Some operators require e.g. `ValEx(TlaStr(_))` values
+ * specifically, instead of just `Str`-typed values (for instance keys in
+ * [[typecomp.ScopedBuilder.rec the record constructor]]). Other operators are variadic, but require the number of
+ * arguments to be of a specific parity (e.g. [[typecomp.ScopedBuilder.funDefMixed funDefMixed]]), or require a positive
+ * number of arguments (e.g. [[typecomp.ScopedBuilder.enumSet enumSet]]). Example:
+ * {{{
+ *   val ex: TlaEx = tla.enumSet() // throws IllegalArgumentException
+ * }}}
+ * would throw an [[IllegalArgumentException]], as [[typecomp.ScopedBuilder.enumSet enumSet]] requires a strictly
+ * positive number of arguments ([[typecomp.ScopedBuilder.emptySet emptySet]] should have been used in this case). </li>
+ *
+ * <li> [[typecomp.TBuilderScopeException TBuilderScopeException]]: This exception indicates that a variable is used
+ * with different types in the same scope, or that shadowing has occurred. Example:
+ * {{{
+ *   // x > 1 /\ x /= "1"
+ *   // throws a TBuilderScopeException
+ *   val ex: TlaEx =
+ *     tla.and(
+ *       tla.ge(
+ *         tla.name("x", IntT1),
+ *         tla.int(1),
+ *       ),
+ *       tla.neql(
+ *         tla.name("x", StrT1),
+ *         tla.str("1")
+ *       )
+ *     )
+ * }}}
+ * In this case, `x` appears as both `Int` and `Str` typed within the scope of `ex`, which is disallowed. </li>
+ *
+ * </ul>
+ *
+ * '''IMPORTANT: EXCEPTIONS WILL ONLY BE THROWN AFTER A `TBuilderInstruction => TlaEx` CONVERSION (IMPLICIT OR
+ * EXPLICIT). INVOKING BUILDER METHODS WITHOUT CONVERTING WILL NOT THROW THE ABOVE EXCEPTIONS.'''
+ */
 package object types {
 
   import at.forsyte.apalache.tla.typecomp._
 
   /**
-   * A short-hand to the instance of the typed builder, so one can easily construct expressions. For example:
+   * A short-hand to the instance of the typed builder, so one can easily construct expressions.
    *
-   * {{{
-   * import at.forsyte.apalache.tla.types._
-   *
-   * test("tla builder succeeds") {
-   *   val ex: TlaEx = tla.plus(tla.int(2), tla.int(3))
-   *   assert(ex.isInstanceOf[OperEx])
-   * }
-   * }}}
-   *
-   * Alternatively, you can import all of its methods via:
+   * You can import all of its methods via:
    *
```

Comment:
Should go below, but I can't mark those (unchanged lines)

```suggestion
   * @see [[types]] for usage instructions.
   * @see [[ScopedBuilder]] for implementation details.
```